proto/public/uber/cadence/api/v1/service_visibility.proto

```diff
@@ -0,0 +1,140 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+syntax = "proto3";
+
+package uber.cadence.api.v1;
+
+option go_package = "apiv1";
+option java_multiple_files = true;
+option java_outer_classname = "ApiProto";
+option java_package = "com.uber.cadence.api.v1";
+
+import "uber/cadence/api/v1/visibility.proto";
+import "uber/cadence/api/v1/workflow.proto";
+
+service VisibilityAPI {
+  // ListWorkflowExecutions is a visibility API to list workflow executions in a specific domain.
+  rpc ListWorkflowExecutions(ListWorkflowExecutionsRequest) returns (ListWorkflowExecutionsResponse);
+
+  // ListOpenWorkflowExecutions is a visibility API to list the open executions in a specific domain.
+  rpc ListOpenWorkflowExecutions(ListOpenWorkflowExecutionsRequest) returns (ListOpenWorkflowExecutionsResponse);
+
+  // ListClosedWorkflowExecutions is a visibility API to list the closed executions in a specific domain.
+  rpc ListClosedWorkflowExecutions(ListClosedWorkflowExecutionsRequest) returns (ListClosedWorkflowExecutionsResponse);
+
+  // ListArchivedWorkflowExecutions is a visibility API to list archived workflow executions in a specific domain.
+  rpc ListArchivedWorkflowExecutions(ListArchivedWorkflowExecutionsRequest) returns (ListArchivedWorkflowExecutionsResponse);
+
+  // ScanWorkflowExecutions is a visibility API to list large amount of workflow executions in a specific domain without order.
+  rpc ScanWorkflowExecutions(ScanWorkflowExecutionsRequest) returns (ScanWorkflowExecutionsResponse);
+
+  // CountWorkflowExecutions is a visibility API to count of workflow executions in a specific domain.
+  rpc CountWorkflowExecutions(CountWorkflowExecutionsRequest) returns (CountWorkflowExecutionsResponse);
+
+  // GetSearchAttributes is a visibility API to get all legal keys that could be used in list APIs.
+  rpc GetSearchAttributes(GetSearchAttributesRequest) returns (GetSearchAttributesResponse);
+}
+
+
+message ListWorkflowExecutionsRequest {
+  string domain = 1;
+  int32 page_size = 2;
+  bytes next_page_token = 3;
+  string query = 4;
+}
+
+message ListWorkflowExecutionsResponse {
+  repeated WorkflowExecutionInfo executions = 1;
+  bytes next_page_token = 2;
+}
+
+message ListOpenWorkflowExecutionsRequest {
+  string domain = 1;
+  int32 page_size = 2;
+  bytes next_page_token = 3;
+  StartTimeFilter start_time_filter = 4;
+  oneof filters {
+    WorkflowExecutionFilter execution_filter = 5;
+    WorkflowTypeFilter type_filter = 6;
+  }
+}
+
+message ListOpenWorkflowExecutionsResponse {
+  repeated WorkflowExecutionInfo executions = 1;
+  bytes next_page_token = 2;
+}
+
+message ListClosedWorkflowExecutionsRequest {
+  string domain = 1;
+  int32 page_size = 2;
+  bytes next_page_token = 3;
+  StartTimeFilter start_time_filter = 4;
+  oneof filters {
+    WorkflowExecutionFilter execution_filter = 5;
+    WorkflowTypeFilter type_filter = 6;
+    StatusFilter status_filter = 7;
+  }
```

Comment:
IMO sorta weird that you can't search for both type and status.  (execution is pretty specific, maybe not necessary to combine with others... except maybe status?)
is there a technical issue preventing it on the backend?

Reply:
It goes all the way down to persistence https://github.com/uber/cadence/blob/master/common/persistence/visibilityInterfaces.go#L181. TBH I'm not sure why this was design that way. But if want to change this, it should be a separate PR.